python/ray/tune/tests/test_tune_restore.py

```diff
@@ -643,6 +644,27 @@ def training_func(config):
             tune.run(training_func)
 
 
+@pytest.mark.parametrize(
+    "trial_config", [{}, {"attr": 4}, {"nested": {"key": "value"}}]
+)
+def test_trial_restore(trial_config):
+    metrics = {"metric1": 4, "nested2": {"metric3": 6}}
+    metrics["config"] = trial_config
+
+    trial = Trial(trainable_name="stub", config=trial_config, stub=True)
+    trial.update_last_result(metrics)
+
+    checkpoint = _TrackedCheckpoint(
+        dir_or_data="no_data",
+        storage_mode=CheckpointStorage.PERSISTENT,
+        metrics=metrics,
+    )
+
+    trial.restoring_from = checkpoint
+    trial.on_restore()
```

Comment:
I'm thinking we can similarly remove `Trial.saving_to` while we're at it.

1. Keep `is_saving` as a flag. OR remove (see below)
2. Create the `_TrackedCheckpoint` when the saving future resolves rather than on start.

### `Trial.is_saving`

This seems to be the only functional use of `is_saving`:

https://github.com/ray-project/ray/blob/5759c41754a60cceb2d457b959fd1601b5281332/python/ray/tune/execution/trial_runner.py#L810

Can we just queue the decision regardless -- then only execute the decision in the queue if the trial has no other running future?

Reply:
`trial.is_restoring` works similar to `is_saving` and is used in two places.

I'm open to refactoring this logic, but it feels like we should do this in a follow-up.

Reply:
Yes, sounds good!